crypt.go

```diff
@@ -550,6 +559,114 @@ func ActivateVolumeWithRecoveryKey(volumeName, sourceDevicePath string, keyReade
 	return activateWithRecoveryKey(volumeName, sourceDevicePath, keyReader, options.Tries, RecoveryKeyUsageReasonRequested, activateOptions)
 }
 
+// ActivationData corresponds to some data added to the user keyring by one of the ActivateVolume functions.
+type ActivationData interface{}
+
+// RecoveryActivationData is added to the user keyring when a recovery key is used to activate a volume.
+type RecoveryActivationData struct {
+	Key    RecoveryKey
+	Reason RecoveryKeyUsageReason
+}
+
+// GetActivationDataFromKernel retrieves data that was added to the user keyring by ActivateVolumeWithTPMSealedKey or
+// ActivateVolumeWithRecoveryKey for the specified source block device. The block device path must match the path passed to one of
+// the ActivateVolume functions. The type of data returned is dependent on how the volume was activated - see the documentation for
+// each function, If no data is found for the specified device, a ErrNoActivationData error is returned.
+func GetActivationDataFromKernel(sourceDevicePath string) (ActivationData, error) {
+	var userKeys []int
+
+	sz, err := unix.KeyctlBuffer(unix.KEYCTL_READ, userKeyring, nil, 0)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot determine size of user keyring payload: %w", err)
+	}
+
+	for {
+		payload := make([]byte, sz)
+		n, err := unix.KeyctlBuffer(unix.KEYCTL_READ, userKeyring, payload, 0)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot read user keyring payload: %w", err)
+		}
+
+		if n <= sz {
+			payload = payload[:n]
+
+			for len(payload) > 0 {
+				userKeys = append(userKeys, int(binary.LittleEndian.Uint32(payload)))
+				payload = payload[4:]
+			}
+			break
+		}
+
+		sz = n
+	}
+
+	re := regexp.MustCompile(`^user;[[:digit:]]+;[[:digit:]]+;[[:xdigit:]]+;secboot:([^\?]+)\??(.*)`)
```

Comment:
if we let the caller control the prefix as we should, this should at least take that in. It can default to "" => "secboot"